fe/fe-core/src/main/java/com/starrocks/system/NodeSelector.java

```diff
@@ -0,0 +1,340 @@
+                Backend backend = systemInfoService.getBackend(backendId);
+                if (backend != null && !checkedHosts.contains(backend.getHost()) &&
+                        !checkedLocs.contains(backend.getSingleLevelLocationKV())) {
+                    locBackendWithDiffHostLocNum++;
+                    checkedHosts.add(backend.getHost());
+                    checkedLocs.add(backend.getSingleLevelLocationKV());
+                }
+            }
+        }
+
+        return  locBackendWithDiffHostLocNum;
+    }
+
+    /**
+     * Choose nodes(currently only backend supported) matching the location requirement.
+     * This is a **best-effort** choosing, if the location requirement cannot be met, for example,
+     * user has specified 3-replica table and "rack:*" as location property, but we only have
+     * 2 different rack tags, the creation won't fail, and the replicas will scatter on the 2 racks
+     * as much as possible.
+     */
+    private List<Long> chooseNodesMatchLocReq(int nodeNum, Multimap<String, String> locReq,
+                                              final List<ComputeNode> srcNodes) {
+        // convert the Multimap to a flatten list so that we can randomly choose backend later
+        // Backend id list which matches the location requirement, classified by location,
+        // e.g. transform from {"rack:rack1": [1, 2, 3], "rack:rack2": [4, 5]} to [[1, 2, 3], [4, 5]].
+        List<List<Long>> locBackendIdList = new ArrayList<>();
+        int locBackendWithDiffHostLocNum = getLocationMatchedBackendIdList(
+                locBackendIdList, srcNodes, locReq, systemInfoService);
+
+        List<Long> nodeIds = Lists.newArrayList();
+        // We cannot find enough nodes match the location requirement,
+        // return an empty list and try to choose nodes ignoring location requirement again.
+        if (locBackendWithDiffHostLocNum < nodeNum) {
+            return nodeIds;
+        }
+
+        // At least we can choose `nodeNum` backends from different hosts (randomly).
+        int startIdx = locBackendRandSelector.nextInt(Integer.MAX_VALUE) % locBackendIdList.size();
+        Set<String> checkedHosts = Sets.newHashSet();
+        while (nodeIds.size() < nodeNum) {
+            List<Long> backendIdList = locBackendIdList.get(startIdx);
+            if (backendIdList.isEmpty()) {
+                startIdx = (startIdx + 1) % locBackendIdList.size();
+                continue;
+            }
+            int idx = locBackendRandSelector.nextInt(Integer.MAX_VALUE) % backendIdList.size();
+            Long backendId = backendIdList.get(idx);
+            Backend backend = systemInfoService.getBackend(backendId);
+            if (backend != null && !nodeIds.contains(backendId) && !checkedHosts.contains(backend.getHost())) {
+                nodeIds.add(backendId);
+                checkedHosts.add(backend.getHost());
+                backendIdList.remove(idx);
+            }
```

Comment:
How about iterating all the backend in backendIdList to find a backend, so that the replicas can be located on different tags as much as possible.

Reply:
For current design, if the location requirement cannot be met, we will ignore the location requirement completely, won't scatter the replica on location matched backend as much as possible, because achieving that will complicate the balancer logic too much. Comments has been modified to match the design.